### ISSUE TITLE
Feature Request: static types when editing

### ISSUE THREAD

**Issue by yoiang** [mac-s-g/react-json-view#163]
Hello! I'm unclear if this relates to #124, I had trouble understanding the ask there so I apologize if this is a duplicate!

When editing a field with a primitive type, regardless of the type `onEdit` passes a result whose `new_value` is a string, rather than maintaining the previous type. It would be great if editing maintained the original type!

_Edit:_
My mistake, I just noticed the two sets of controls. Is there a way to only show the type safe set?

**Comment by yoiang** [comment]
My mistake, I just noticed the two sets of controls. Is there a way to only show the type save set?

**Comment by mac-s-g** [comment]
when editing a value, RJV attempts to parse and detect primitives.  When it does, you'll get an additional submit button indicating the detection.

If you submit with `ctrl+enter` and there's a detected primitive, the primitive will be submitted by default.

I hope this answers your question.  If not, don't hesitate to re-open this issue.

**Comment by yoiang** [comment]
I apologize, I should have been clearer! It’d be great to have a component
prop option to only allow type valid values from edits.

**Comment by mac-s-g** [comment]
would string be a valid type?  that's the default type for all inputs currently.

**Comment by mac-s-g** [comment]
oh, maybe you're thinking there could be a prop (say `editType`) that dictates the input type.  
so if `editType = "integer"`, then you would see a number input and only numbers would be valid inputs.

is that what you're thinking?

**Comment by yoiang** [comment]
I was thinking for pre-populated JSON where a member has a value of a certain type already specified having an prop that makes sure a user can only update that member with a value of the same type that it already is.

```json
{
  iAmANumber: 1234,
  i_am_a_string: "asdf" 
}
```
In this example with this prop enabled editing `iAmANumber` would only show the UI for accepting a valid **number** type where `i_am_a_string` would only show the UI for accepting a valid **string** type. The type safe acceptance UI are already available to _react-json-view_, it would be a matter of hiding the UI that stores the value as a string. 

It's unclear what this means for `null` values, however, and would require something extra for `array` types, either some smarts on the component's part or something similar to JSON schema. But for at least one usage I have in mind even not allowing altering `null` or `array` types (or adding new members to an `object`) would be a gem. It would essentially make _react-json-view_ into a very easy to use and quick to implement data form.

**Comment by yoiang** [comment]
Would you be open to a PR with just this prop, one that hid the UI to save a value as a **string** and only showed the type safe UI when editing?

**Comment by mac-s-g** [comment]
yes, but how will you know the type without a json schema?

**Comment by yoiang** [comment]
I'm assuming that the component's one set of controls currently determine the type of a member based on the value of that member. 

When a JSON Schema isn't present the prop would:

* only allow editing values that already have a value (and thus their type can be determined)
OR
* will also allow assigning **null** with a value and then that value becomes the determining type. This second one seems more complicated, there'd have to be a way to specify string vs anything else.

**Comment by mac-s-g** [comment]
so with option 1, if you edit an integer value, you would only be able to input an integer.  is that right?

**Comment by yoiang** [comment]
Exactly, the UI, as it currently does for the type safe input, would only
let you save a valid number. Integer vs float I feel less strongly about
since they aren't basic Javascript types

**Comment by mac-s-g** [comment]
yea that makes sense.  maybe call the prop `staticTypes` or something?
I'd be okay with adding a feature like that.
If you're comfortable posting a PR, I can definitely review.  I can also take it on, but I might not be able to get to it for a while.

I'm gonna re-open this issue as a `staticTypes` feature request.

**Comment by yoiang** [comment]
I have some free time coming up, I'll take a stab at it 😊

Static types as in the prop ensures that the type of the member does not change?

**Comment by mac-s-g** [comment]
Cool. Yea, that's what I was thinking.

**Comment by yoiang** [comment]
🤘

**Comment by mrbfrank** [comment]
@yoiang @mac-s-g Hi I have a similar use case wishing to lock data types, did we come to any conclusion here with impementing `staticTypes` or `JSON Schema`?

I do think `JSON Schema` makes the most sense, for instance if you have a null value.. you'd need some way to declare the appropriate type when adding a value.

**Comment by yogiindia** [comment]
Its great user is provided with 2 submit options. But if a developer wants to force validation, then, in that case, single submit option must be there. There must be an option to turn off 2 submit option and instead provide only one.
